crates/red_knot_python_semantic/src/semantic_index/builder.rs

```diff
@@ -725,8 +737,7 @@ where
                     // which is invalid syntax. However, it's still pretty obvious here that the user
                     // *wanted* `e` to be bound, so we should still create a definition here nonetheless.
                     if let Some(symbol_name) = symbol_name {
-                        let symbol = self
-                            .add_or_update_symbol(symbol_name.id.clone(), SymbolFlags::IS_DEFINED);
```

Comment:
The initial motivation was that I didn't want the knowledge of which Definitions are a Binding to be spread across various places in the builder, which is error prone; I wanted it consolidated in one place, in the impl I added to `DefinitionKind`. But since we need a symbol in order to create a definition, if we make Definition responsible for knowing whether a symbol is a binding, that means we can't set `SymbolFlags::BOUND` when creating the symbol. But this actually made the API cleaner IMO; `add_symbol` is now responsible only for adding the symbol (it doesn't update anything if called a second time for the same symbol), and we have separate methods for marking the symbol as bound and/or used. And we can always mark a symbol as bound in just one place, in `add_definition`, using the Definition's own knowledge of whether it is a binding or not.